### PR TITLE
feat: Avm/res/network/express route circuit 

### LINKED ISSUE #285: Fix upload-module-metadata
The pipeline fails due to a 404. Need to find out the root cause and fix it.

### PR DESCRIPTION
**Pull request by arnoldna**
## Description

Migration of CARML Bicep module for Express-Route-Circuits to AVM closes #285

I[avm.res.network.express-route-circuit](https://github.com/arnoldna/bicep-registry-modules/actions/workflows/avm.res.network.express-route-circuit.yml)

### CODE REVIEW

File: avm/res/network/express-route-circuit/main.bicep [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    properties: {
+      peeringType: peeringType
+      sharedKey: sharedKey
+      peerASN: peerASN
+      primaryPeerAddressPrefix: primaryPeerAddressPrefix
+      secondaryPeerAddressPrefix: secondaryPeerAddressPrefix
+      vlanId: vlanId
+    }
+  }
+]
```
Comment (AlexanderSehr) [inline avm/res/network/express-route-circuit/main.bicep:100]: As this cariable is only used in one location, I'd suggest to move it into the actual resource deployment below. You should be able to replace it 1:1

Comment (AlexanderSehr) [reply avm/res/network/express-route-circuit/main.bicep:100]: Note, if changed may require a regeneration of the `main.json` file

File: avm/res/network/express-route-circuit/main.bicep [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+output resourceId string = expressRouteCircuits.id
+
+@description('The resource group the express route curcuit was deployed into.')
+output resourceGroupName string = resourceGroup().name
+
+@description('The name of express route curcuit.')
+output name string = expressRouteCircuits.name
+
+@description('The service key of the express route circuit.')
+output serviceKey string = reference(expressRouteCircuits.id, '2021-02-01').serviceKey
```
Comment (AlexanderSehr) [inline avm/res/network/express-route-circuit/main.bicep:213]:
```suggestion
output serviceKey string = expressRouteCircuits.properties.serviceKey
```

Should be simpler

File: avm/res/network/express-route-circuit/tests/e2e/waf-aligned/main.test.bicep [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      kind: 'CanNotDelete'
+      name: 'myCustomLockName'
+    }
+    roleAssignments: [
+      {
+        roleDefinitionIdOrName: 'Reader'
+        principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+        principalType: 'ServicePrincipal'
+      }
+    ]
```
Comment (AlexanderSehr) [inline avm/res/network/express-route-circuit/tests/e2e/waf-aligned/main.test.bicep:94]:
```suggestion
```

Should not be required to be WAF-aligned. Note: If removed will need a refresh of the readme.

File: avm/res/network/express-route-circuit/main.bicep [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        properties: {
+          peeringType: peeringType
+          sharedKey: sharedKey
+          peerASN: peerASN
+          primaryPeerAddressPrefix: primaryPeerAddressPrefix
+          secondaryPeerAddressPrefix: secondaryPeerAddressPrefix
+          vlanId: vlanId
+        }
+      }
+    ]
```
Comment (AlexanderSehr) [inline avm/res/network/express-route-circuit/main.bicep:146]: Not exactly what I had in mind. The parameter should still be optional.  I was more thinking along the lines of:

```suggestion
    peerings: peering ? [
      {
        name: peeringType
        properties: {
          peeringType: peeringType
          sharedKey: sharedKey
          peerASN: peerASN
          primaryPeerAddressPrefix: primaryPeerAddressPrefix
          secondaryPeerAddressPrefix: secondaryPeerAddressPrefix
          vlanId: vlanId
        }
      }
    ] : null
```

Comment (AlexanderSehr) [reply avm/res/network/express-route-circuit/main.bicep:146]: Please add the parameter back and undo the related changes ;)

File: avm/res/network/express-route-circuit/tests/e2e/max/main.test.bicep [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+// ============== //
+@batchSize(1)
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    name: '${namePrefix}${serviceShort}001'
+    location: location
+    bandwidthInMbps: 50
+    peeringLocation: 'Amsterdam'
```
Comment (AlexanderSehr) [inline avm/res/network/express-route-circuit/tests/e2e/max/main.test.bicep:68]: I guess we cannot test the `peering` parameter, right? I assume that would require a working endpoint... Just wondering.

Comment (arnoldna) [reply avm/res/network/express-route-circuit/tests/e2e/max/main.test.bicep:68]: Correct. We don't have a way to test that.

File: avm/res/network/load-balancer/tests/e2e/internal/main.test.bicep [hunk 6]
```diff
@@ -1,5 +1,8 @@
 targetScope = 'subscription'
 
+metadata name = 'Using internal load balancer parameter'
+metadata description = 'This instance deploys the module with the minimum set of required parameters to deploy an internal load balancer.'
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/tests/e2e/internal/main.test.bicep:4]: That's not entirely true as you also deploy diagnostics etc. If you want to use this text I'd recommend removing anything that isn't needed for an internal load balancer to ensure a consumer isn't confused on what's actually required

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]

File: avm/res/network/load-balancer/tests/e2e/waf-aligned/main.test.bicep [hunk 7]
```diff
@@ -1,5 +1,8 @@
 targetScope = 'subscription'
 
+metadata name = 'WAF-aligned'
+metadata description = 'This instance deploys the module in alignment with the best-practices of the Well-Architected Framework.'
```
Comment (AlexanderSehr) [inline avm/res/network/load-balancer/tests/e2e/waf-aligned/main.test.bicep:4]: Note, adding the metadata requires a rerun of the Readme generation. I'd advice to do that and then run the pipeline with only the static tests enabled to check if all passes

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]

**Review by AlexanderSehr — APPROVED** [review]

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**